train_llama3.cu

```diff
@@ -883,29 +925,39 @@ void gpt2_backward_and_reduce(GPT2 *model, int* inputs, const int* targets, int
         floatX* l_ln2 = (model->recompute < 2) ? acts.ln2 + l * B * T * C : acts.lnf;
         float* l_ln2_mean = acts.ln2_mean + l * B * T;
         float* l_ln2_rstd = acts.ln2_rstd + l * B * T;
-        floatX* l_fch_pre_gelu = acts.fch + l * B * T * 4*C;
-        floatX* l_fch_gelu = (model->recompute < 1) ? acts.fch_gelu + l * B * T * 4*C : acts.fch_gelu;
+        floatX* l_fch_pre_act = acts.fch + l * B * T * 4*C;
+        floatX* l_fch_pre_act2 = acts.fch2 + l * B * T * 4*C;
+        floatX* l_fch_act = (model->recompute < 1) ? acts.fch_act + l * B * T * 4*C : acts.fch_act;
         // get the pointers of the gradients of the activations for this layer
         // notice that there is no l *, because we just have a single copy, and keep
         // re-using this memory in every Transformer block as we calculate backward pass
 
         floatX* dl_bt4c = (floatX*)model->acts.scratch_bt4c;
+        floatX* dl_bt4c2 = (floatX*)model->acts.scratch_bt4c2;
 
         // start the backward pass for this layer
         if(model->recompute >= 1) {
-            // recompute >= 1 means we recompute gelu. in this case,
-            // l_fch_gelu is just a buffer, so re-compute the gelu from l_fch here
-            gelu_forward(l_fch_gelu, l_fch_pre_gelu, B*T*4*C, main_stream);
+            if (strcmp(model->act_func, "gelu") == 0) {
```

Comment:
would it make sense to do a call to a single activation_forward() call, and rename gelu.cuh to activation.cuh since it now also does SwiGLU?

Assuming we want to keep this flexible at all, but it feels like experimenting with different activation functions is an easy thing that'd be worth doing a little bit more in the future (e.g. loss impact & GPU throughput of RELU and RELU squared for example).

Right now we'd just pass all the arguments of swiglu_forward() since gelu_forward()'s arguments are a subset (that won't scale if adding more complicated activation functions but I think that's probably OK/still useful?)